### PR TITLE
Fix models being broken after changing column order

### LINKED ISSUE #29951: Changing column order on model view then almost saving without preview breaks model with custom columns
**Describe the bug**
- Changing column order of model with custom column breaks model

**Logs**
image

**To Reproduce**
Steps to reproduce the behavior (if you can reproduce the bug using the Sample Database, we will find the issue faster):
1. Create and save a question from Accounts table in Sample Database with all columns.
2. Turn the question into a model
3. Edit query definition, add custom column "Seating" with `CASE([Seats] = 1, "Single", [Seats] = 2, "Double", "Other")`
4. Save question (without clicking preview)
5. Drag new "Seating" column to a different position (in model view, not in model query editor)
6. Click Save, but don't save, click Cancel
7. Click preview

**Expected behavior**
- Expect to save

**Screenshots/videos**
image

**Information about your Metabase Installation:**

- Metabase version: 46.1 and master (as of 2023-04-10)

**Severity**
- moderate, can cause loss of unsaved changes

### PR DESCRIPTION
**Pull request by ranquild**
Fixes https://github.com/metabase/metabase/issues/29951

The problem was that we used the original question (the model) when re-running the query when saving changes to the model and returning to the chill mode. But the chill mode expects to work with an-hoc question based on the model, and there was an issue where we used `results_metadata` from the query being run for the wrong question. This PR solves the issue by making sure we re-run the query with the correct question.

The "correct question" here means what we get from `getQuestion` selector. It depends on the `queryBuilderMode`, so it's important to set it to `view` (the chill mode) before executing the query after leaving the dataset editor. I've left more comments in the PR about the change.

How to verify:
- Either follow the steps from the original issue closely or run the cypress test

Screenshot 2023-04-13 at 17 51 24

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/30073)

### CODE REVIEW

File: frontend/src/metabase/query_builder/actions/core/core.js [hunk 1]
```diff
@@ -225,7 +225,7 @@ export const apiUpdateQuestion = (question, { rerunQuery } = {}) => {
     const originalQuestion = getOriginalQuestion(getState());
     question = question || getQuestion(getState());
 
-    rerunQuery = rerunQuery || getIsResultDirty(getState());
+    rerunQuery = rerunQuery ?? getIsResultDirty(getState());
```
Comment (ranquild) [inline frontend/src/metabase/query_builder/actions/core/core.js:228]: To not re-run the model query when it's updated in this action. The problem is that `runQuestionQuery` depends on the `queryBuilderMode`, and we need to set it first before running the query. So the dataset editor calls this action with `rerunQuery: false`, sets the query builder mode, and then runs it manually.

File: frontend/src/metabase/query_builder/actions/core/core.js [hunk 2]
```diff
@@ -264,9 +264,10 @@ export const apiUpdateQuestion = (question, { rerunQuery } = {}) => {
 
     dispatch({ type: API_UPDATE_QUESTION, payload: updatedQuestion.card() });
 
+    const metadataOptions = { reload: question.isDataset() };
```
Comment (ranquild) [inline frontend/src/metabase/query_builder/actions/core/core.js:267]: `rerunQuery` will be `false` when updating the model, but we still need to load metadata when the question is updated.

File: frontend/src/metabase/query_builder/containers/QueryBuilder.jsx [hunk 3]
```diff
@@ -275,7 +275,7 @@ function QueryBuilder(props) {
   );
 
   const handleSave = useCallback(
-    async (updatedQuestion, { rerunQuery = false } = {}) => {
+    async (updatedQuestion, { rerunQuery } = {}) => {
```
Comment (ranquild) [inline frontend/src/metabase/query_builder/containers/QueryBuilder.jsx:278]: Removed the default value to distinguish between `undefined` (for default behavior) and `false` (where we explicitly ask not to run the query)

File: frontend/src/metabase/query_builder/components/DatasetEditor/DatasetEditor.jsx [hunk 4]
```diff
@@ -306,13 +307,14 @@ function DatasetEditor(props) {
     if (canBeDataset && isBrandNewDataset) {
       onOpenModal(MODAL_TYPES.SAVE);
     } else if (canBeDataset) {
-      await onSave(dataset);
-      setQueryBuilderMode("view");
+      await onSave(dataset, { rerunQuery: false });
+      await setQueryBuilderMode("view");
+      runQuestionQuery();
```
Comment (ranquild) [inline frontend/src/metabase/query_builder/components/DatasetEditor/DatasetEditor.jsx:312]: `apiUpdateQuestion` doesn't wait for the query results and we shouldn't do it here, too

**Comment by WiNloSt** [comment]
@ranquild 
I couldn't reproduce the issue. I think I probably do something wrong as I have tried to follow the steps laid out in https://github.com/metabase/metabase/issues/29951, but after step 4
> 4. Save question (without clicking preview)

there are 2 scenarios.
1. If I start everything from the query editor without even viewing the model via run button or click the **visualize** button, I can't modify the question's column without either clicking the **visualize** button or run a query button on the table view since there's no gear option on the bottom left without running a query.
    Screenshot 2023-04-14 at 12 55 49 PM


2. If I have run a query on the model since step 1 or 2, I won't be able to change the seating column position without adding it to the query (clicking the plus button), which will run the query and modify the order after that also doesn't encounter any error.
Screenshot 2023-04-14 at 12 53 49 PM

Also, I couldn't reproduce the bug using the repro test from this PR running against `master` 072c0f391a30183cf0a3e5d2e8c9260db32cf7c8
 code. 

If I do that it will fail at this step that expects the request `dataset` to run
Screenshot 2023-04-14 at 12 44 35 PM

If I removed that the test will pass 
image
Screenshot 2023-04-14 at 12 45 08 PM

That's the reason I think the repro doesn't correctly capture the bug.

**Comment by ranquild** [comment]
@WiNloSt The thing is that in the current (not fixed) version a wrong endpoint is used to load model dataset in the query builder after updating the model, so the repro should modified if you want to see the original error. Here is the diff for that:

```
diff --git a/e2e/test/scenarios/models/reproductions/29951-model-editor-results-metadata.cy.spec.js b/e2e/test/scenarios/models/reproductions/29951-model-editor-results-metadata.cy.spec.js
index 3bb0c22ee1..094be31539 100644
--- a/e2e/test/scenarios/models/reproductions/29951-model-editor-results-metadata.cy.spec.js
+++ b/e2e/test/scenarios/models/reproductions/29951-model-editor-results-metadata.cy.spec.js
@@ -24,6 +24,7 @@ describe("issue 29951", () => {
     restore();
     cy.signInAsAdmin();
     cy.intercept("POST", "/api/dataset").as("dataset");
+    cy.intercept("POST", "/api/card/*/query").as("cardQuery");
     cy.intercept("PUT", "/api/card/*").as("updateCard");
   });
 
@@ -36,7 +37,7 @@ describe("issue 29951", () => {
     removeExpression("CC2");
     cy.findByRole("button", { name: "Save changes" }).click();
     cy.wait("@updateCard");
-    cy.wait("@dataset");
+    cy.wait("@cardQuery");
 
     dragColumn(0, 100);
     cy.findAllByRole("button", { name: "Get Answer" }).first().click();
diff --git a/frontend/src/metabase/query_builder/actions/core/core.js b/frontend/src/metabase/query_builder/actions/core/core.js
index 83b3f2f6ac..5c1520d7de 100644
--- a/frontend/src/metabase/query_builder/actions/core/core.js
+++ b/frontend/src/metabase/query_builder/actions/core/core.js
@@ -225,7 +225,7 @@ export const apiUpdateQuestion = (question, { rerunQuery } = {}) => {
     const originalQuestion = getOriginalQuestion(getState());
     question = question || getQuestion(getState());
 
-    rerunQuery = rerunQuery ?? getIsResultDirty(getState());
+    rerunQuery = rerunQuery || getIsResultDirty(getState());
 
     // Needed for persisting visualization columns for pulses/alerts, see #6749
     const series = getTransformedSeries(getState());
diff --git a/frontend/src/metabase/query_builder/components/DatasetEditor/DatasetEditor.jsx b/frontend/src/metabase/query_builder/components/DatasetEditor/DatasetEditor.jsx
index 6362fd1006..ec1aedbdd3 100644
--- a/frontend/src/metabase/query_builder/components/DatasetEditor/DatasetEditor.jsx
+++ b/frontend/src/metabase/query_builder/components/DatasetEditor/DatasetEditor.jsx
@@ -182,7 +182,6 @@ function DatasetEditor(props) {
     height,
     isDirty: isModelQueryDirty,
     setQueryBuilderMode,
-    runQuestionQuery,
     setDatasetEditorTab,
     setFieldMetadata,
     onCancelDatasetChanges,
@@ -307,14 +306,14 @@ function DatasetEditor(props) {
     if (canBeDataset && isBrandNewDataset) {
       onOpenModal(MODAL_TYPES.SAVE);
     } else if (canBeDataset) {
-      await onSave(dataset, { rerunQuery: false });
+      await onSave(dataset);
       await setQueryBuilderMode("view");
-      runQuestionQuery();
+      // runQuestionQuery();
     } else {
       onOpenModal(MODAL_TYPES.CAN_NOT_CREATE_MODEL);
       throw new Error(t`Variables in models aren't supported yet`);
     }
-  }, [dataset, onSave, setQueryBuilderMode, runQuestionQuery, onOpenModal]);
+  }, [dataset, onSave, setQueryBuilderMode, onOpenModal]);
 
   const handleColumnSelect = useCallback(
     column => {

```

When you comment `cy.wait`, you drag the column in the dataset editor and not the chill mode, that's why it seems like the repro isn't working but it does. Please take a look at this.

**Comment by WiNloSt** [comment]
@ranquild Thank you that works perfectly, I can reproduce the problem now. The key is to use **Edit query definition** from the entity menu, I was clicking **Show editor** button and edit the query there. 👍

**Review by WiNloSt — APPROVED** [review]
I can reproduce the issue with the diff provided by @ranquild . This works great.

> Reviewed 3 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/30073#-:-NT-J9Ho5NqITBNBSxEc:bo0p8sa)* status: 3 of 4 files reviewed, 4 unresolved discussions (waiting on @kulyk, @NevRA, and @ranquild)

___
*[`frontend/src/metabase/query_builder/actions/core/core.js` line 228 at r1](https://reviewable.io/reviews/metabase/metabase/30073#gh-1165658908:-NT-9sgK8dcKJX3Fca-T:b44jn0a) ([raw file](https://github.com/metabase/metabase/blob/b0e23d35fbfdd0f98c8d0a5683b60844d7d1760b/frontend/src/metabase/query_builder/actions/core/core.js#L228)):*
<details><summary><i>Previously, ranquild (Alexander Polyankin) wrote…</i></summary><blockquote>

To not re-run the model query when it's updated in this action. The problem is that `runQuestionQuery` depends on the `queryBuilderMode`, and we need to set it first before running the query. So the dataset editor calls this action with `rerunQuery: false`, sets the query builder mode, and then runs it manually.
</blockquote></details>

Not to block the PR, but do you think in general we should avoid using a boolean with 3 states (null, true, false)? I find when dealing with booleans with 3 states, it feels very awkward, and sometimes catches me off guard since perceptually I would only expect booleans to be either true or false (or undefined) all the time.
___
*[`frontend/src/metabase/query_builder/actions/core/core.js` line 267 at r1](https://reviewable.io/reviews/metabase/metabase/30073#gh-1165660143:-NT-DM1T5v5Pj1SS9iUU:bn0kqx1) ([raw file](https://github.com/metabase/metabase/blob/b0e23d35fbfdd0f98c8d0a5683b60844d7d1760b/frontend/src/metabase/query_builder/actions/core/core.js#L267)):*
<details><summary><i>Previously, ranquild (Alexander Polyankin) wrote…</i></summary><blockquote>

`rerunQuery` will be `false` when updating the model, but we still need to load metadata when the question is updated. It's cached so there is no harm in calling this method more than needed.
</blockquote></details>

I'm not sure what's cached in this case. What I've tried is to just copy this line a bunch of times.
e.g. 
```
    await dispatch(loadMetadataForCard(question.card(), metadataOptions));
    await dispatch(loadMetadataForCard(question.card(), metadataOptions));
    await dispatch(loadMetadataForCard(question.card(), metadataOptions));
    await dispatch(loadMetadataForCard(question.card(), metadataOptions));
```

I can see that there are multiple calls to `/api/database/:database-id/schema` and `api/table/:table-id/query_metadata` In this particular repro, after saving the query definition there are 3 calls for each line of `await dispatch(loadMetadataForCard(question.card(), metadataOptions));`

I just want to check if I understand what you're explaining correctly. Or is it another cache you're talking about?

**Pull request merged**